Add sequence sessionizer.

nnyx7 opened the pull request:
The provided sequence sessionizer works like an interface for specific
sequence sessionizers.
It resolves issue #956.

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,152 @@
+
+from __future__ import unicode_literals
+
+#from timesketch.lib.analyzers import manager
```
Remove commented line

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,152 @@
+
+    Attributes:
+        event_seq_name: The name of the event sequence.
+        event_seq: List of dictionaries of attributes describing the events
```
```suggestion
        event_seq: List of dictionaries of attributes describing the events.
```

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,152 @@
+        event_seq: List of dictionaries of attributes describing the events
+        event_storage: List of events.
+        num_event_to_find: Number that shows which event from the sequence
+        should be matched.
```
Please indent with 4 spaces when doing a newline in the `Args` section
```suggestion
            should be matched.
```

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,152 @@
+        num_event_to_find: Number that shows which event from the sequence
+        should be matched.
+        recording: Shows if currently are finded and stored events that match
+        the specified sequence of events.
```
```suggestion
            the specified sequence of events.
```

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,152 @@
+        recording: Shows if currently are finded and stored events that match
+        the specified sequence of events.
+        return_fields: List of name of event attributes, should be specified
+        in the inheriting sessionizers. It's mandatory to contains 'timestamp'.
```
```suggestion
            in the inheriting sessionizers. It must contain 'timestamp'.
```

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,152 @@
+    session_num = 0
+
+    def run(self):
+        """Entry point for the analyzer. Allocates each event between the
```
You can insert a newline before `Allocates` to have a neater docstring:

```
"""Entry point for the analyzer

Allocates each event...

Returns:
    String containing...
```

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,152 @@
+        if self.event_seq is None or self.event_seq == []:
+            raise RuntimeError('No event_seq provided.')
+        if 'timestamp' not in self.return_fields:
+            raise RuntimeError('No \'timestamp\' in return_fields.')
```
Alternatively, you can use double quotes around `"timestamp"` to avoid escaping single quotes

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,152 @@
+
+        self.sketch.add_view('Session view', self.NAME, query_string=self.query)
+
+        return ('Sessionizing completed, number of {0} session created:'
```
```suggestion
        return ('Sessionizing completed, number of {0:s} session created:'
```

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,152 @@
+        event.commit()
+
+    def process_event(self, event):
+        """Process event depending on if the event is significant for the
```
Same here, try to have a docstring with a brief one-line description, and insert details in a paragraph underneath.

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,152 @@
+            self.event_storage.append(event)
+
+    def flush_events(self, drop=False):
+        """Flush the events in the event_storage.
```
Maybe write something more descriptive here? "Annotates or clears events in event_storage and resets session recording state"

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,152 @@
+        Args:
+            event: Event to compare with the one in the event sequence.
+
+        Returns: If event matches the event to search for in the event
```
Returns should be on a newline:

```
Returns:
    If event matches...
```

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer_test.py:
```diff
@@ -0,0 +1,391 @@
+            message = sessionizer.run()
+            self.assertEqual(
+                message,
+                'Sessionizing completed, number of {0} session created: 1'.
```
```suggestion
                'Sessionizing completed, number of {0:s} session created: 1'.
```

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer_test.py:
```diff
@@ -0,0 +1,391 @@
+
+def _create_mock_event(event_id, quantity, attributes, time_diffs=None):
+    """
+    Returns an instance of Event, based on the MockDataStore event_dict
```
Does this really return an instance of Event? Or rather generates a list of Event objects?

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer_test.py:
```diff
@@ -0,0 +1,391 @@
+
+    Args:
+        event_id: Desired ID for the Event.
+        quantity: The number of Events to be generated
```
```suggestion
        quantity: The number of Events to be generated.
```

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer_test.py:
```diff
@@ -0,0 +1,391 @@
+        event_id: Desired ID for the Event.
+        quantity: The number of Events to be generated
+        attributes: A list of dictionaries with attributes for the generated
+        events.
```
```suggestion
            events.
```

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer_test.py:
```diff
@@ -0,0 +1,391 @@
+                format(sessionizer.event_seq_name))
+
+            ds = MockDataStore('test', 0)
+            event1 = (ds.get_event('test_index', '0', stored_events=True))
```
Why the extra parentheses around the `ds.get_event()` call? I see this in many places in the code

nnyx7 commented:
I changed the function create_eventObj to be more flexible.
There was a bug. The event_template object was not deep copied and changed every time which was not problem before because of the way the function create_eventObj was written, but I had to fix it for the currect version.

nnyx7 commented:
The SequenceSessionizerSketchPlugin is dependent from the added additional changes in commit 5233e50f3a0aea0488554aadb2b7dc56e7876a43. Those changes are the same like the ones in the Expert sessionizer pull request.

nnyx7 commented:
A session_type is new attribute with default value that SequenceSessionizerSketchPlugin inherits from SessionizerSketchPlugin. It behaved the same like the attribute event_seq_name so I replace event_seq_name with session_type.

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,158 @@
+        if self.event_seq is None or self.event_seq == []:
+            raise RuntimeError('No event_seq provided.')
+        # If return_fields in none, then all attributes are provided.
+        if self.return_fields is not None:
```
Didn't we discuss building a minimal list of `return_fields` depending on the dicts in `event_seq`? It's OK to leave it for a future PR but in this case please add a `# TODO` comment.

nnyx7 replied:
My logic there is that if someone chosed to have all of the fields, he probably needs all of them.
If someone didn't specified the return_fields in his sessionizer, then return_fields remains ['timestamp'] and I am going to fill the rest of the fields that are needed. And a field is needed if event from event_seq has it.
If someone specified return_fields but not with the right attributes for some reason, I am not removing the not needed ones, I am just adding the attributes that are needed but missing. So in that case return_fields can end up with not used attributes.
But the another side of situation is if the person needs to access other attributes too, then to removing them is unexpected behaviour.
In the first situation I am adding extra complexity but making sure that the code will work as expected in the sequence_sessionizer. In the second situation, I can break someone's logic.

nnyx7 replied:
What do you think it's better to be done?

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,158 @@
+            if 'timestamp' not in self.return_fields:
+                raise RuntimeError('No "timestamp" in return_fields.')
+            if len(self.return_fields) <= 1:
+                raise RuntimeError('No return_fields provided.')
```
Either change the condition to `<= 0` or change the error message to `"Not enough return_fields provided"`

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,158 @@
+            last_timestamp = first_event.source.get('timestamp')
+            self.process_event(first_event)
+
+            for event in events:
```
You can probably change this loop to

```python
            last_timestamp = None
            for event in events:
                curr_timestamp = event.source.get('timestamp')
                if last_timestamp and (curr_timestamp - last_timestamp > self.max_time_diff_micros):
                    self.flush_events(drop=True)
                self.process_event(event)
                last_timestamp = curr_timestamp
```

and skip the bloc where you call `next(events)`. Or am I missing something?

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,158 @@
+                             self.NAME,
+                             query_string=self.query)
+
+        return ('Sessionizing completed, number of {0:s} session created:'
```
```suggestion
        return ('Sessionizing completed, number of {0:s} sessions created:'
```

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,158 @@
+        """Process event depending on if the event is significant for the
+        searched event sequence.
+
+        Event is significant if it matches the current
```
Please try to reach the max number of words in a line before breaking.

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,158 @@
+        drop and resets session recording state.
+
+        Args:
+            drop: Indicator if the stored events should not be commited.
```
This sentence is a bit unclear.

```suggestion
            drop: If True, the stored events will not be committed to the database.
```

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,158 @@
+
+        Returns:
+            If event matches the event to search for in the event
+        sequence.
```
```suggestion
              sequence.
```

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,158 @@
+                return False
+        return True
+
+    def get_query_string(self):
```
If this function is meant to be overriden by other classes, please specify it in the docstring.

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer.py:
```diff
@@ -0,0 +1,158 @@
+        Returns:
+            Query string for Elasticsearch.
+        """
+        query_string = self.session_type + ':*'
```
```suggestion
        return self.session_type + ':*'
```

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer_test.py:
```diff
@@ -0,0 +1,430 @@
+        """Test session_type is not None."""
+        index = 'test_index'
+        sketch_id = 1
+        self.assertRaises(RuntimeError, sessionizer=\
```
It's better to write it this way:

```python
self.assertRaises(
    RuntimeError, 
    sessionizer=NoneSessionTypeSequenceSessionizer(index, sketch_id))
```

Same for the other assertions in the tests.

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer_test.py:
```diff
@@ -0,0 +1,430 @@
+            message = sessionizer.run()
+            self.assertEqual(
+                message,
+                'Sessionizing completed, number of {0:s} session created: 1'.
```
Quick note: if you fix the string in the previous comments, you'll have to adjust this string too.

```suggestion
                'Sessionizing completed, number of {0:s} sessions created: 1'.
```

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer_test.py:
```diff
@@ -0,0 +1,430 @@
+    @mock.patch('timesketch.lib.analyzers.interface.ElasticsearchDataStore',
+                MockDataStore)
+    def test_multiple_sessions(self):
+        """Test multiple sessions are finded and allocated correctly."""
```
```suggestion
        """Test multiple sessions are found and allocated correctly."""
```

tomchop commented on timesketch/lib/analyzers/sequence_sessionizer_test.py:
```diff
@@ -0,0 +1,430 @@
+            ds = MockDataStore('test', 0)
+            for i in range(0, 100):
+                event = ds.get_event('test_index', str(i), stored_events=True)
+                self.assertEqual(event['_source'][sessionizer.session_type], 1)
```
In tests, I would rather be explicit here and use the actual hardcoded string instead of the variable `sessionizer.session_type` when possible.
